### 2. More Querying Data - COUNT, DISTINCT, GROUPBY, MAX, MIN SUM, AVG
This workbook expands on workbook 1, giving you more tools and keywords to query your data.

This includes use agregation functions to summarise data and get statistics about it, such as sum, average, maximum and minimum.

###### Keywords Covered - COUNT, DISTINCT, GROUPBY, MAX, MIN, SUM, AVG

In [1]:
import sqlite3
import pandas as pd

# Connect to database
data_conn = sqlite3.connect("pokemon.db")

#### COUNT

Use: Returning the number of rows that match a criteria.

*Tip: Use in conjunction with other keywords such as DISTINCT, WHERE and GROUPBY for deeper insights.*

In the last workbook we used the len() python function to count the rows of data. We can get the same result using the COUNT SQL keyword.

In [2]:
# Using len() function in python
data = data_conn.execute("SELECT * FROM pokemon").fetchall()
print(f"Rows in data: {len(data)}")

poke_count = data_conn.execute("SELECT COUNT (*) FROM pokemon").fetchall()
print("Rows in data using COUNT function:", poke_count[0][0])

Rows in data: 1025
Rows in data using COUNT function: 1025


#### DISTINCT

Use: Removing duplicates and showing only unique values.

Below we will use the DISTINCT keyword to list the types of Pokemon, and then the COUNT keyword to count them.

In [3]:
d_types = data_conn.execute("SELECT DISTINCT type1 FROM pokemon").fetchall()
print(pd.DataFrame(d_types))

count_d_types = data_conn.execute("SELECT COUNT (DISTINCT  type1) FROM pokemon").fetchall()
print("Pokemon Types:",count_d_types[0][0])

           0
0        bug
1       dark
2     dragon
3   electric
4      fairy
5   fighting
6       fire
7     flying
8      ghost
9      grass
10    ground
11       ice
12    normal
13    poison
14   psychic
15      rock
16     steel
17     water
Pokemon Types: 18


#### GROUPBY 

Use: Organises data into groups based on one or more column chosen to group by.

*Tip: Use with aggregate functions like MIN, MAX, COUNT, SUM or AVG.*

Below we will use GROUP BY with the COUNT function to count how many of each Pokemon type there are.

In [4]:
d_types = data_conn.execute("SELECT type1, COUNT(name) FROM pokemon GROUP BY type1").fetchall()
print(pd.DataFrame(d_types))

           0    1
0        bug   83
1       dark   45
2     dragon   37
3   electric   59
4      fairy   29
5   fighting   40
6       fire   66
7     flying    9
8      ghost   35
9      grass  103
10    ground   40
11       ice   31
12    normal  118
13    poison   42
14   psychic   60
15      rock   58
16     steel   36
17     water  134


#### MAX

Use: Find the maximum value of a column.

Below we will find the tallest Pokemon. Note how three columns are being SELECTed.

In [5]:
tall_pkmn = data_conn.execute("SELECT name, type1, MAX(height) FROM pokemon").fetchall()
print(pd.DataFrame(tall_pkmn))

           0       1     2
0  Eternatus  poison  20.0


#### Min

Use: Find the minimum value of a column.

Below we will find the shortest Pokemon of each type by use MIN in conjunction with the GROUP BY.

In [6]:
short_kings = data_conn.execute("SELECT name, type1, MIN(height) FROM pokemon GROUP BY type1").fetchall()
print(pd.DataFrame(short_kings))

                  0         1    2
0            Joltik       bug  0.1
1          Purrloin      dark  0.4
2             Goomy    dragon  0.3
3            Tynamo  electric  0.2
4           Flabebe     fairy  0.1
5            Mankey  fighting  0.5
6         Scorbunny      fire  0.3
7          Rookidee    flying  0.2
8          Sinistea     ghost  0.1
9      Poltchageist     grass  0.1
10          Diglett    ground  0.2
11             Snom       ice  0.3
12          Azurill    normal  0.2
13         Shroodle    poison  0.2
14          Cosmoem   psychic  0.1
15          Carbink      rock  0.3
16           Klefki     steel  0.2
17  Wishiwashi Solo     water  0.2


#### SUM

Use: Find the total value of a column.

Below we will sum the total of all the Pokemon IDs from 1 to 1025.

We can check this as the total of any number from 1 to n is n(n+1)/2.

So the answer shoudl be 1025(1025 + 1) / 2 = 525,825

In [7]:
id_total = data_conn.execute("SELECT SUM(id) FROM pokemon").fetchall()
print("Total:", id_total[0][0])

Total: 525825


#### AVG

Use: Calculate the mean average of a column.

*Tip: use with GROUP BY to create useful statistics.*

Below we will calculate the average wight of each Pokemon type group.

In [8]:
avg_weight = data_conn.execute("SELECT type1, AVG(height) FROM pokemon GROUP BY type1").fetchall()
print(pd.DataFrame(avg_weight))

           0         1
0        bug  0.877108
1       dark  1.320000
2     dragon  1.800000
3   electric  1.096610
4      fairy  0.841379
5   fighting  1.310000
6       fire  1.201515
7     flying  1.177778
8      ghost  1.148571
9      grass  0.909709
10    ground  1.390000
11       ice  1.316129
12    normal  1.018644
13    poison  1.626190
14   psychic  1.050000
15      rock  1.365517
16     steel  1.794444
17     water  1.388806
